lang/src/org/partiql/lang/planner/transforms/AstToLogicalVisitorTransform.kt

```diff
@@ -81,6 +86,62 @@ internal class AstToLogicalVisitorTransform(
         return expr
     }
 
+    private fun generateUniqueWindowName(node: PartiqlAst.Expr.CallWindow, index: Int): String {
+        val uniqueId = "\$__partiql_window_function_$index"
+        windowFunctionUniqueIdMap[node] = uniqueId
+        return uniqueId
+    }
+
+    private fun getUniqueWindowName(node: PartiqlAst.Expr.CallWindow): String {
+        return windowFunctionUniqueIdMap[node] ?: error("no such window function registered")
+    }
+
+    /**
+     * If the Select Clause contains window function, create a corresponding Window operator.
+     */
+    private fun extractWindowFunction(node: PartiqlAst.Expr.Select, algebra: PartiqlLogical.Bexpr): PartiqlLogical.Bexpr {
+        // check to see if there is any window call in the select list
+        val windowExpressions = when (val project = node.project) {
+            is PartiqlAst.Projection.ProjectValue -> return algebra
+            is PartiqlAst.Projection.ProjectList -> {
+                project.projectItems.asSequence().filterNot {
+                    it is PartiqlAst.ProjectItem.ProjectAll
+                }.filter {
+                    (it as PartiqlAst.ProjectItem.ProjectExpr).expr is PartiqlAst.Expr.CallWindow
+                }.map {
+                    (((it as PartiqlAst.ProjectItem.ProjectExpr).expr) as PartiqlAst.Expr.CallWindow)
+                }
+            }
```

Comment:
Currently trying to get the semantics down, but is there any way where the ProjectList expression will ever hold anything else?

I know the the changes in `PartiQL.g4` aren't part of this PR -- but they're relevant -- `windowFunction` is an `exprPrimary` and therefore can be expressed as a sub-expression. So, a query like:
```sql
SELECT 5 + LAG(t.co) OVER() FROM t
```

Could theoretically be possible, unless a visitor pass throws an exception on this -- but if there wasn't some sort of pass on this, then this would be allowed, and the lines referenced above would not be able to extract the window function. You'd unfortunately need to create a visitor to grab them.

Reply:
BTW the above is allowed in Postgres too -- just checked to make sure

Reply:
Spot on! Working on this now!

Reply:
Came up with a Visitor Transform that is inspired by your group by transformer. See the new commit.